rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1059 @@
+-> qux
+
+[
+  some
+  flags
+]
+++ ( # <- The operator is on a new line, but parenthesis/brackets/braces can start on the same line
+  foo
+)
+++ optionals condition [ # <- As are multiline functions applications
+  more
+  items
+]
+++
+  runCommand name # <- But only functions whose last argument can start on the same line
+    ''
+      echo hi
+    ''
+    test
+
+# We can use a more compact form as long as all inbetween lines are indented
+bar // {
+  x = 10;
+  y = 20;
+} // baz
+
+# Bad, we can't use the more compact form because an intermediate line is not indented.
+{
+  x = 10;
+  y = 20;
+} // foo // {
+  z = 30;
+  w = 40;
+}
+```
+
+### if
+
+**Desciption**
+
+- `if` and `else` keywords always start a line, the if and else bodies are indented.
+- If the condition does not fit onto one line, then it will start on the next line with indentation, and `then` will be on the start of the line following the condition.
+- `else if` chains are treated as one long sequence, with no indentation creep on each step.
+- Only simple `if` statement can be single-line, no `else if` chains.
+
+**Examples**
+
+```nix
+#1
+if builtins.length matches != 0 then
+  { inherit path matches; }
+else if path == /. then
+  [
+    1
+    2
+  ]
+else
+  go (dirOf path);
+
+#2
+if
+  matches != null
+  && builtins.length matches != 0
+then
+  { inherit path matches; }
+else if path == /. then
+  null
+else
+  go (dirOf path);
+```
+
+**Rationale and alternatives**
+
+- Attribute sets and lists could start on the same line as the if keywords, saving an indentation level on their body:
+  ```nix  
+  #1a
+  if builtins.length matches != 0 then {
+    inherit path matches;
+  } else if path == /. then [
+    1
+    2
+  ] else
+    go (dirOf path);
+  ```
+  - This results in inconsistent vertical start of the keywords, making the structure harder to follow
+- Have the `then` on the start of the next line, directly followed by the if body:
+  ```nix
+  #1b
+  if builtins.length matches != 0
+    then { inherit path matches; }
+  else if path == /.
+    then [
+      1
+      2
+    ]
+  else go (dirOf path);
+
+  #1c
+  if builtins.length matches != 0
+  then { inherit path matches; }
+  else if path == /.
+  then [
+    1
+    2
+  ]
+  else go (dirOf path);
+  ```
+
+### with, assert
+
+**Description**
+
+- The body after the statement starts on a new line, without indentation.
+- There may be exceptions for common idioms, in which the body already starts on the same line.
+
+**Examples**
+
+```nix
+with pkgs;
+assert foo == bar; {
+  meta.maintainers =
+    with lib.maintainers; [
+      some
+      people
+    ];
+}
+```
+
+### let
+
+**Description**
+
+- Let bindings are *always* multiline.
+- The "let" part is indented one level, but not the "in" part.
+- Each item in the "let" part is indented and starts on its own line.
+  For more details, see the [binders section](#binders).
+- The "in" part starts on a new line.
```

Comment:
Is there a reason we can't smooth that down for expressions that fit in a line, so we can have something like

```nix
let
  # lots of stuff
in result
```

Reply:
The motivation behind always having a newline is for easier editing and better diffs. E.g. say you have

```nix
let
  foo = 10;
in foo
```

And you want to insert a function argument (or another `let in` or an `assert`/`with`), you'd have to go inbetween `in` and `foo` to do that, and the diff would be

```diff
 let
   foo = 10;
-in foo
+in
+bar:
+foo
```

In the current proposal it would be just

```diff
 let
   foo = 10;
 in
+bar:
 foo
```

However, we need to consider trade-offs, I have the feeling that it might be worthwhile to allow no newline after `in`.

Reply:
Minimal diffs are generally good, but as you say one has to consider it in context and weigh diff readability against code readability.

Reply:
The reason why I'd be against the compact form here is that this kind of looks weird in combination with most statements but simple identifiers (`in if`, `in foo: bar`. So for consistency's sake it's easier to always have a newline. But if you think that simple identifiers look too weird on their own line, we could special case just that

Reply:
Really not a fan of not indenting the inner expression after the `in` but still putting it on a new line. To me this feels very wrong, like formatting the following in C:

```c
return
some_method_call(
  foo,
  bar
);
```

instead of

```c
return some_method_call(
  foo,
  bar
);
```

IMHO always attaching the `in` to the first line of the inner expression looks and reads better:

```nix
let
  x = 1;
in 2 * x

let
  x = 1;
in [
  x
  (x + 1)
  (x + 2)
]

let
  foo = bar;
in stdenv.mkDerivation {
  ...
}
```

And generally I think the same for attaching the `let` to the end of the last line of the outer expression for the same reason:

```nix
{
  stdenv,
}: let
  foo = ...;
in stdenv.mkDerivation {
  ...
}
```

Reply:
From what I've seen in @infinisil's code contributions to Nixpkgs, the motivation behind the unindented newline pattern could be to encourage using `let` blocks to declare the provenance of symbols at the top of a file (the antithesis to `with pkgs; with lib;`). You'd usually leave an empty line between that and the business logic. Since it's a homogeneous rule you can repeat that pattern in nested scopes. I think it's quite nice.

Reply:
In the meeting today we reached general agreement that it's okay to always have a newline after `in`

Reply:
> Can we consider forcing a newline after `in`? The examples above read somewhat poorly to me. I think `in stdenv.mkDerivation {` reads best

I'm not sure what you're arguing for here, you can't both force a newline after `in` and then show an example where that's not the case :)

Reply:
Right. What I meant to say is, every example that isn't `in stdenv.mkDerivation {` read badly to me. So I'm thinking if forcing a newline after `in` would be a net positive, considering that there's only one example of it being a little worse.

Reply:
You are mistaken about the proposed style here. In combination with other rules, you get the following instead:

```nix
  config =
    let
      foo = "bar";
    in
    mkIf (this.enable) {
      services.foo.bar = foo;
    };
```

(As you already said, that is a tradeoff between an additional indentation level for the content and keeping the top-level indent tidy)

Reply:
We discussed this in the meeting again and added the mentioned alternatives to the RFC now: 6a6efcd2b5588ba662004b48314204628487f551